provisionersdk/logger.go

```diff
@@ -0,0 +1,71 @@
+package provisionersdk
+
+import (
+	"bufio"
+	"io"
+
+	"github.com/coder/coder/provisionersdk/proto"
+)
+
+type Logger interface {
+	Log(l *proto.Log) error
+}
+
+type ProvisionStream interface {
+	Send(*proto.Provision_Response) error
+}
+
+type ParseStream interface {
+	Send(response *proto.Parse_Response) error
+}
+
+func NewProvisionLogger(s ProvisionStream) Logger {
+	return provisionLogger{s}
+}
+
+func NewParseLogger(s ParseStream) Logger {
+	return parseLogger{s}
+}
+
+type provisionLogger struct {
+	stream ProvisionStream
+}
+
+func (p provisionLogger) Log(l *proto.Log) error {
+	return p.stream.Send(&proto.Provision_Response{
+		Type: &proto.Provision_Response_Log{
+			Log: l,
+		},
+	})
+}
+
+type parseLogger struct {
+	stream ParseStream
+}
+
+func (p parseLogger) Log(l *proto.Log) error {
+	return p.stream.Send(&proto.Parse_Response{
+		Type: &proto.Parse_Response_Log{
+			Log: l,
+		},
+	})
+}
+
+func LogWriter(logger Logger, level proto.LogLevel) (io.WriteCloser, <-chan any) {
```

Comment:
This function seems very specific to how we are running terraform and moves some of the granularity out of terraform, whilst creating a dependency provisionersdk. Some thoughts:

- We're assuming log messages will be newline delimited (scanner)
- We cannot control log level on a per-log line basis (may be relevant for future provisioners)
- This function creates an `io.Pipe` but never closes r or w, closing w is implicitly up to the caller (the done channel makes it a bit less clear we're supposed to close)
- In acutal uses, we're relying on `exec.Command` to close w (i.e. there's a lot of implicitness going on)